libnavui-util/src/main/java/com/mapbox/navigation/ui/utils/internal/resource/ResourceLoader.kt

```diff
@@ -0,0 +1,82 @@
+package com.mapbox.navigation.ui.utils.internal.resource
+
+import com.mapbox.bindgen.Expected
+import com.mapbox.common.ResourceLoadError
+import com.mapbox.common.ResourceLoadProgress
+import com.mapbox.common.ResourceLoadResult
+
+/**
+ * ResourceLoader is responsible for downloading and storage of any Nav SDK BLOB assets.
+ */
+interface ResourceLoader {
+
+    /**
+     * Asynchronously load the resource.
+     *
+     * If the requested resource is not available in device storage a download will be queued.
+     * The [ResourceLoadRequest.flags] and [ResourceLoadRequest.networkRestriction] can be used to
+     * control loading behaviour.
+     *
+     * If the downloaded data causes the disk quota to be exceeded, a quota enforcement operation is
+     * scheduled.
+     *
+     * @param request Resource Load Request
+     * @param callback Called to notify of load operation progress
+     *
+     * @return requestId Request ID that can be used to cancel load operation.
+     */
+    fun load(request: ResourceLoadRequest, callback: ResourceLoadCallback): Long
```

Comment:
The cache can be "queried" by passing ResourceLoadRequest with `ResourceLoadFlags.SKIP_DATA_LOADING` and `NetworkRestriction.DISALLOW_ALL`. The result will only contain the resource metadata, without loading the data from the disk. If the resource is  not in the cache the result will have the Unsatisfied error type.